bend_gcode.py

```diff
@@ -5,135 +5,91 @@
-        height = SplineLookupTable[i]
-        if height >= Zheight:
-            return i * DISCRETIZATION_LENGTH
-    print("Error! Spline not defined high enough!")
-
-def createSplineLookupTable():
-    heightSteps = np.arange(DISCRETIZATION_LENGTH, SPLINE_Z[-1], DISCRETIZATION_LENGTH)
-    for i in range(len(heightSteps)):
-        height = heightSteps[i]
-        SplineLookupTable.append(SplineLookupTable[i] + np.sqrt((SPLINE(height)-SPLINE(height-DISCRETIZATION_LENGTH))**2 + DISCRETIZATION_LENGTH**2))
-    
-
-
-lastPosition = Point2D(0, 0)
-currentZ = 0.0
-lastZ = 0.0
-currentLayer = 0
-relativeMode = False
-createSplineLookupTable()
-
-with open(INPUT_FILE_NAME, "r") as gcodeFile, open(OUTPUT_FILE_NAME, "w+") as outputFile:
-        for currentLine in gcodeFile:
+@dataclass
+class Point2D:
+    x: float
+    y: float
+
+@dataclass
+class GCodeLine:
+    x: float
+    y: float
+    z: float
+    e: float
+    f: int
+
+    gcode_format: ClassVar[re.Pattern] = (
+        re.compile('(?i)^[gG][0-3](?:\s'
+                   '+x(?P<x>-?[0-9.]{1,15})|\s'
+                   '+y(?P<y>-?[0-9.]{1,15})|\s'
+                   '+z(?P<z>-?[0-9.]{1,15})|\s'
+                   '+e(?P<e>-?[0-9.]{1,15})|\s'
+                   '+f(?P<f>-?[0-9.]{1,15}))*')
+    )
+
+    @classmethod
+    def from_gcode(cls, line: str):
+        line_entries = cls.gcode_format.match(line)
+        if line_entries:
+            return cls(*line_entries.groups())
+
+
+def main(in_file, out_file, spline, layer_height, warning_angle,
+         xy_precision=4, z_precision=3, e_precision=5):
+
+    lastPosition = Point2D(0, 0)
+    currentZ = 0.0
+    lastZ = 0.0
+    currentLayer = 0
+    relativeMode = False
+
+    with open(in_file, "r") as gcodeFile, open(out_file, "w+") as outputFile:
+        def writeLine(G, X, Y, Z, E=None, F=None):
+            output = f'G{G} X{X:.{xy_precision}f} Y{Y:.{xy_precision}f} Z{Z:.{z_precision}f}'
+            if E is not None:
+                output += f" E{E:.{e_precision}f}"
+            if F is not None:
+                output += f" F{F}"
+            outputFile.write(output + '\n')
+
+        print('Processing: (press CTRL+C to abort)')
+        start = perf_counter()
+        for index, currentLine in enumerate(gcodeFile):
+            if index % 1000 == 0: # track progress
+                print(f"[{timedelta(seconds=perf_counter()-start)}]: line {index}\r", end='')
             if currentLine[0] == ";":   #if NOT a comment
                 outputFile.write(currentLine)
                 continue
-            if currentLine.find("G91 ") != -1:   #filter relative commands
+            if currentLine.startswith("G91"):    #filter relative commands
                 relativeMode = True
                 outputFile.write(currentLine)
                 continue
-            if currentLine.find("G90 ") != -1:   #set absolute mode
+            if currentLine.startswith("G90 "):   #set absolute mode
```

Comment:
You removed the trailing space on L72 but left it in L76. Is there a reason for this or was it an oversight?

I feel that both `str().startswith` and `str().find` are brittle and should be replaced with a dedicated function.

Reply:
> You removed the trailing space on L72 but left it in L76. Is there a reason for this or was it an oversight?

I think I actually added an extra space on 72 so the comments would line up between 72 and 76, since they already lined up (which can aid readability). Once again, not sure if an intentional commenting style from Stefan, or just the number of spaces that happened to end up on those lines.

> I feel that both `str().startswith` and `str().find` are brittle and should be replaced with a dedicated function.

`find` could conceivably be incorrect if G90 appears in a comment. `startswith` should be correct, although could be marginally more robust with an `lstrip()` first. I chose to avoid that because it's then less clear, and I figured the input gcode is expected to come from a slicer, and I don't believe they ever put spaces before lines, but perhaps that's not the case (maybe some use indenting or something?).

Do you have any particular preference as to what should be there?

Reply:
More generally, I expect a bunch of the logic throughout this function could be cleaned up somewhat, I just left it alone beyond where I thought something was error-prone, or needed to change due to the spline refactoring. In my profiling the logic itself wasn't an obvious performance throttle, so I figured my time would be more efficiently spent elsewhere :-)